qubes/storage/callback.py

```diff
@@ -0,0 +1,442 @@
+            return None
+        else:
+            return self._cb_impl.included_in(app)
+
+    @property
+    def size(self):
+        if self._cb_requires_init:
+            return None
+        else:
+            return self._cb_impl.size
+
+    @property
+    def usage(self):
+        if self._cb_requires_init:
+            return None
+        else:
+            return self._cb_impl.usage
+
+    #remaining method & attribute delegation ("delegation pattern")
+    #Convention: The methods of this object have priority over the delegated object's methods. All attributes are
+    #           passed to the delegated object unless their name starts with '_cb_'.
+    def __getattr__(self, name):
+        #NOTE: This method is only called when an attribute cannot be resolved locally (not part of the instance,
+        #       not part of the class tree). It is also called for methods that cannot be resolved.
+        return getattr(self._cb_impl, name)
+
+    def __setattr__(self, name, value):
+        #NOTE: This method is called on every attribute assignment.
+        if name.startswith('_cb_'):
+            super().__setattr__(name, value)
+        else:
+            setattr(self._cb_impl, name, value)
+
+    def __delattr__(self, name):
+        if name.startswith('_cb_'):
+            super().__delattr__(name)
+        else:
+            delattr(self._cb_impl, name)
+
+class CallbackVolume:
+    ''' Proxy volume adding callback functionality to other volumes.
+
+        Required to support the `on_sinit` callback for late storage initialization.
+
+        **Important for Developers**: Even though instances of this class behave exactly as `qubes.storage.Volume` instances,
+                                    they are no such instances (e.g. `assert isinstance(obj, qubes.storage.Volume)` will fail).
+    '''
+
+    def __init__(self, pool, impl):
+        '''Constructor.
+        :param pool: `CallbackPool` of this volume
+        :param impl: `qubes.storage.Volume` object to wrap
+        '''
+        assert isinstance(impl, qubes.storage.Volume), 'impl must be a qubes.storage.Volume instance. Found a %s instance.' % impl.__class__
+        assert isinstance(pool, CallbackPool), 'pool must use a qubes.storage.CallbackPool instance. Found a %s instance.' % pool.__class__
+        self._cb_pool = pool
+        self._cb_impl = impl
+
+    def _assertInitialized(self, **kwargs):
+        return self._cb_pool._assertInitialized(**kwargs)
+
+    def _callback(self, cb, cb_args=[], **kwargs):
+        vol_args = [ *cb_args, self.name, self.vid ]
+        return self._cb_pool._callback(cb, cb_args=vol_args, **kwargs)
+
+    def create(self):
+        self._assertInitialized()
+        self._callback('on_volume_create')
+        return self._cb_impl.create()
+
+    def remove(self):
+        self._assertInitialized()
+        ret = self._cb_impl.remove()
+        self._callback('on_volume_remove')
+        return ret
```

Comment:
Here you're actually doing the callback after the backend's `remove` method has started (not finished!) if it's a coroutine, i.e. with ThinVolume and ReflinkVolume. As a quick and dirty fix, I think you could check if `ret` is a coroutine object and if so, splice in `ret = asyncio.get_event_loop().run_until_complete(ret)`. But the real fix would be to make your method a coroutine too and use `qubes.utils.coro_maybe`. Then you'd also need an async `_callback` variant. Ah, function coloring hell...

Same in `CallbackVolume.stop` and `CallbackVolume.import_data_end`, and theoretically in `CallbackPool.destroy`.

Reply:
No, you definitely can't do `asyncio.get_event_loop().run_until_complete()`. It is not legal operation if event loop is running elsewhere (it is). This needs to be a proper coroutine.